### PR TITLE
BlockParam* classes refactoring

### PR DESCRIPTION
**Pull request by bkueng**
The existing BlockParam classes have some issues, which IMO prevent the use of them by default.
- The biggest issues is the memory overhead: each instance has a linked list pointer, a pointer to the vtable (because it contains virtual methods), and a `param_t` handle (16 bits on NuttX, but because of the memory layout in that specific case, it will still use 32 bits). While the actual data we care about is only 32 bits.
  Since we have >1000 params by now and tendency increasing, this will account for several kb's of (wasted) RAM, if those classes are used everywhere.
- if the parent naming prefix is used, grepping the src tree for a param name will not find all usages. I already ran into problems because of this.
- the parent naming prefix is optional, which makes the API confusing
- the namespace name `control` is not appropriate (it's not only used for controllers anymore, so we should rename it). Any ideas?

Otherwise, the existing API is ok, which is why I largely tried to keep it, and only change the deficiencies.

### Proposed solution
Since in the majority of the cases, we know which param we want already at compile-time, I decided to use an enum instead of a string to specify the parameter. This has benefits in terms of runtime (no `param_find()` required), and memory (no need to store the handle). In addition it's a template parameter, so that the type can be checked at compile-time.

This is how it's used: `BlockParamFloat<px4::params::MPC_MANTHR_MAX> _manual_thr_max`

#### Automatic Parameter updates
One thing missing from that is the automatic parameter update calls. As it turns out this is more difficult to achieve without overhead, and minimal boilerplate. What I wanted to do is to iterate over all fields of a struct. This requires a language-feature called Reflection, but C++ does not support that. The closest I found is [this](https://github.com/apolukhin/magic_get), which aliases a struct as a `tuple` (using lots of template magic so it's quite involved). A [different solution](http://pfultz2.com/blog/2012/07/31/reflection-in-under-100-lines/) uses macros, and I decided to do something similar: the block param definitions are wrapped in a macro, which automatically generates a (virtual) method that calls `.update()` for all params. This is then called as usual from the `Block` base class on a param update notification.

Final usage example (see 4c7d623 ):
```
	DEFINE_BLOCK_PARAMETERS(
		(BlockParamFloat<px4::params::MPC_MANTHR_MIN>) _manual_thr_min, /**< minimal throttle output when flying in manual mode */
		(BlockParamFloat<px4::params::MPC_MANTHR_MAX>) _manual_thr_max
	)
```

#### Potential programming/user errors
The use of a macro can lead to some usage errors:
- forget to use the macro or use the block param classes outside of the macro.
  Avoided by placing the class definitions into a special namespace, that makes it clear that it should not be used explicitly. In addition, we can add a class definition outside the namespace, that leads to a clear compile error message (using a `static_assert`), that the macro should be used instead.
- In a custom class hierarchy, the child class must call the `updateParams` method of the parent. But C++ does not have a way to generically find the parent class'es name, so the macro cannot do that automatically. Instead, the macro marks the method as `final`, so that a wrong usage will lead to a compile error, forcing the user to use the extended macro, with explicit parent class (however this use-case is rather atypical, and I don't think we have it currently in our code-base).

#### Benefits
- param type statically checked (type-safety w/o runtime overhead)
- no invalid param handles anymore: A used param is guaranteed to be included in the build
- reduced memory usage. The new interface is even better than the typical uses of the C API, since no param handle is stored. A block param really only requires the storage of the param (float or int32)
- reduced runtime overhead: no `param_find()` & linked list operations -> faster bootup
- the default constructor can be used (no additional code in the constructor's initializer list of the SuperBlock required)
- src tree can still be grepped for parameter names

#### Drawbacks
- no dynamic parameter names (the templated enum requires the param name to be known at compile-time). But we only use dynamic strings at a few places, and we can still use the C API for those.
- no arrays possible (BlockParamFloat _mag_params[3]). Currently only used once in ekf2
- the need for a macro, which is not so nice

Note that this only shows the basic working example, and will require additional cleanup.

Sorry, this description got way longer than expected.

Thoughts?

### CODE REVIEW

**Comment by dagar** [comment]
Interesting! I'll spend some time going through the PR, but this definitely sounds like the right general direction.

One immediate thing that comes to mind is to drop "Block" from the name in this version.

**Comment by dagar** [comment]
Given the parameters are defined per module, and used exclusively in that module the vast majority of the time, what if we generated the entire boilerplate per module?

 - struct of BlockParams with appropriate type
 - block of update code
 - possibly tied into the SuperBlock automatically?

In special cases you'd still be able to manually add and access parameters from another module (although I often don't you should).

**Comment by bkueng** [comment]
> Given the parameters are defined per module and used exclusively in that module the vast majority of the time, what if we generated the entire boilerplate per module?

I thought about this, but I think it will become a bit too restrictive. Take navigator for example: it has various Block classes that all have their own parameters. It's nicely modular that way. Similar for the flight tasks: each task will (or can) have it's own parameters.
But of course, if you have an alternative, don't hesitate to bring it up.

**Comment by dagar** [comment]
Mainly brainstorming here, but what if we generated "parameter tables" + update code for each of the parameter definition files. In navigator the different parameters per navigation state are already broken up like this.

Personally I'd rather err on the side of things being restrictive by default as long as there's always an escape hatch. If you write a module that conforms to "best practices" huge portions of error prone manual boilerplate disappear, and you can really focus on the "business logic". Adding a new parameter becomes a trivial exercise. However, we don't want to limit the possibilities, so there's nothing stopping you from ignoring the generated stuff and going your own way, even using the same underlying pieces (BlockParams or old C API) if you wish.

**Comment by dagar** [comment]
We also have this version still laying around. https://github.com/PX4/Firmware/blob/master/src/platforms/px4_parameter.h

**Comment by bkueng** [comment]
> If you write a module that conforms to "best practices" huge portions of error prone manual boilerplate disappear, and you can really focus on the "business logic"

Right, and I think we have that now: using a param requires adding only a single line of code. If you remove even that, I think it becomes too intransparent for a developer (and things like, he/she cannot chose an appropriate variable name). Plus we wanted to move away from the `_params.c` files.

**Comment by dagar** [comment]
> using a param requires adding only a single line of code

2-3 :smirk: 


Don't get me wrong, this is still much better. I share your concerns about transparency. I regularly help people with simple things like adding params and using params, and either way (BlockParam or traditional) there's enough magic going on that it's hard to fully understand without understanding the full picture. I just can't help but wonder if there's a way to skip over even having good boilerplate. 

What I was thinking of was if you just had a const reference to a struct with your parameter subset. Then in the code you could directly access `params.MPC_MANTHR_MIN` or similar. In the rare cases you need to commit, you manually use a BlockParam or the old API. Adding a new param would literally be 1 line.

**Comment by dagar** [comment]
All that being said, I'm in favour of this PR as an updated BlockParam and I'll help get it in if needed. My one nitpicky ask would be if we can try to make it slightly less verbose. Once settled I'd also like to make a concerted effort to use the new/updated API across the majority of the code base.

**Comment by bkueng** [comment]
> 2-3 😏

Alright, depending on how or what you count 😃. I see your point though. Your method is probably doable, but still requires adding a macro to the class (and provide a module-specific struct name, that can be derived from the module name for example) to include the param struct & for the update method. Problems I see are that you don't see the type of the param (it's important to know whether you are working with a float or an int), you lose some generality: params used by multiple modules need to be handled differently (logger and VTOL come to my mind), library code is excluded (might be solvable though), and should we decide to centralize the params with JSON (or something else), it would probably require this to be changed again.
To me, given the gain (removed 1 line per param), does not outweigh the drawbacks.

**Comment by bkueng** [comment]
I did the following updates:
- moved the classes to a better place
- dropped the Block* prefix & `control` namespace
- handling of build dependencies (there might be a somewhat better way?)

### Subscription/Publication classes
I planned to refactor these classes too at a later point, but it's becoming relevant for this PR as well (though I want to do a separate PR for that).
My idea is to make the ORB_ID a template parameter and provide 2 variations: one which contains the topic data, and another that does not. The version without the data basically only contains the file descriptor and allows for more efficient implementations which keep the topic data on the stack, for example like EKF2 is currently structured.
The definition will then look like this:
```
Subscription<parameter_update_s, ORB_ID(parameter_update)> param_update_sub;
SubscriptionWithData<vehicle_gps_position_s, ORB_ID(vehicle_gps_position)> gps_position_sub;
```
Benefits:
- less error prone (topic data type & ORB_ID are right next to each other)
- no separate initializer required
- allows for better type-checking at compile-time
- (much!) easier to extract uorb graph information
- reduced overhead (vtable pointer, orb_metadata is a compile-time param)

Drawback: cannot be stored in a generic list/array. However looking at Flight Tasks for example, all we really need is the topic data and a way to update all subscriptions. This can easily be achieved by a subscription array that uses the C-API internally and returns a pointer to the requested topic data. Alternatively we can still add a non-templated abstract base class if needed.


### Block/SuperBlock classes
Initially I planned to keep those, and just remove the unneeded overhead in the base class (the list of parameters and the name are not needed, and the `dt` is only used in LPE & some of the Block classes themselves). A heavy base class quickly adds up wasting resources (we've seen that before).
In addition to removing the Subscription & Publication lists (not needed in many cases as only the main module contains subs/pubs), this leaves the class pretty much empty.
For these reasons I propose to add a new base class `ModuleParams` that will be used instead to handle the automatic param update propagation. Additional benefits:
- Only a single base class, which avoids confusion between SuperBlock & Block (I've seen them being used wrong in multiple cases). But this might as well just be due to lack of class documentation.
- no automatic `update()` call for all subscriptions/publications. This can be seen as a drawback, but I think it's actually better, because then you get a chance to know which topics got updated. This can be used to reduce computational load and allows more control over when a topic is updated or published.
- incremental refactoring is possible
- only pay for what you use/want: it's more modular, allowing you to inherit only from the things you need.

The existing Block & SuperBlock classes can be reduced eventually to only contain `dt` (which is their main use within controllib).

As an example, I updated the template module: 300af3c

Thoughts/criticism/better ideas?

**Comment by dagar** [comment]
I missed your last update 8 days ago, but I definitely like the ModuleParams base class idea, and in fact wanted to take that idea further. I'll review this in detail. 

On the topic of Subscriptions I have some fundamental questions/ideas I'd like to discuss.

**Comment by bkueng** [comment]
I changed the navigator to use the new Param classes and RAM usage is already reduced by 1 KB.

This is now ready for review and/or merge. I'd like to get this in soon (if people are happy with it), to avoid rebase troubles. The refactoring of the rest can be done in steps.

@dagar I removed some `final` keywords (https://github.com/PX4/Firmware/pull/8789/commits/a440a9813461a6a1295ad2aa74421d960644c743), since I think we should not artificially restrict class hierarchies. This is in accordance to the C++ Core Guidelines: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-final. Of course if you feel otherwise, we can discuss it.

**Comment by bkueng** [comment]
> On the topic of Subscriptions I have some fundamental questions/ideas I'd like to discuss.

Sure, if you want to discuss it directly, we can have a call, or write me on slack (but not today anymore).

File: src/modules/navigator/navigator_vtol_params.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Delay in seconds before applying back transition throttle
+ * Set this to a value greater than 0 to give the motor time to spin down.
+ *
+ * unit s
+ * @min 0
+ * @max 10
+ * @increment 1
+ * @decimal 2
+ * @board px4fmu-v2
```
Comment (dagar) [inline src/modules/navigator/navigator_vtol_params.c]: board? This definitely isn't specific to fmu-v2.

Comment (bkueng) [reply src/modules/navigator/navigator_vtol_params.c]: Look at the commit: https://github.com/PX4/Firmware/pull/8789/commits/0a4f756d1dad5579395347cb748e62262bc8c03d. The goal for now is to fix missing params and we can find a nice solution later.

Comment (dagar) [reply src/modules/navigator/navigator_vtol_params.c]: Yes I saw, but still didn't quite understand. These VTOL params are only used in navigator, so I'd say they need to live in navigator completely.

If we had a bit more control we could only mark them active if a VTOL, but really I'd say this is a clear sign of bad architecture. The better solution would be handling this entirely in VTOL and publishing something, similar to the way FW acceptance radius works now. @RomanBapst 

For now does moving them entirely to navigator solve the problem?

Comment (bkueng) [reply src/modules/navigator/navigator_vtol_params.c]: Turns out `modules/vtol_att_control` was already in v2, and the failing build was due to flash overflow, and not missing params. I removed the file accordingly.

File: cmake/configs/nuttx_mindpx-v2_default.cmake [hunk 2]
```diff
@@ -94,6 +94,7 @@ set(config_module_list
 	modules/attitude_estimator_q
 	modules/position_estimator_inav
 	modules/local_position_estimator
+	modules/landing_target_estimator
```
Comment (RomanBapst) [inline cmake/configs/nuttx_mindpx-v2_default.cmake]: @bkueng Is this explicitly needed for this PR? I'm not proposing to make a separate PR just want to understand.

Comment (bkueng) [reply cmake/configs/nuttx_mindpx-v2_default.cmake]: Yes, there's a parameter dependency from LPE. The commit message contains the details: https://github.com/PX4/Firmware/pull/8789/commits/70db3d2c5582e302a7bd638f28d823870d65dac3

**Comment by jgoppert** [comment]
Lgtm.

**Comment by bkueng** [comment]
Thanks for the reviews
